app/src/main/java/com/github/mik629/android/fundamentals/data/repositories/MoviesRepositoryImpl.kt

```diff
@@ -1,42 +1,70 @@
 package com.github.mik629.android.fundamentals.data.repositories
 
-import com.github.mik629.android.fundamentals.data.mappers.Mapper
-import com.github.mik629.android.fundamentals.data.network.ServerApi
-import com.github.mik629.android.fundamentals.data.network.model.Actor
-import com.github.mik629.android.fundamentals.domain.model.ActorItem
-import com.github.mik629.android.fundamentals.domain.model.GenreItem
-import com.github.mik629.android.fundamentals.domain.model.MovieItem
+import com.github.mik629.android.fundamentals.data.MoviesLoader
+import com.github.mik629.android.fundamentals.data.db.daos.MovieDao
+import com.github.mik629.android.fundamentals.data.db.models.toCrossRef
+import com.github.mik629.android.fundamentals.data.db.models.toEntity
+import com.github.mik629.android.fundamentals.domain.model.Actor
+import com.github.mik629.android.fundamentals.domain.model.Genre
+import com.github.mik629.android.fundamentals.domain.model.Movie
+import com.github.mik629.android.fundamentals.domain.model.MovieDetails
 import com.github.mik629.android.fundamentals.domain.repositories.MoviesRepository
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+import timber.log.Timber
 
 class MoviesRepositoryImpl(
-    private val serverApi: ServerApi,
-    private val actorMapper: Mapper<Actor, ActorItem>
+    private val moviesLoader: MoviesLoader,
+    private val movieDao: MovieDao
 ) : MoviesRepository {
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
 
-    override suspend fun getMovies() = withContext(Dispatchers.IO) {
-
-        val movies = serverApi.getMovieList()
-            .results
-
-        val res = mutableListOf<MovieItem>()
-        for (movie in movies) {
-            val movieDetails = serverApi.getMovieDetails(movie.id)
-            val actors = serverApi.getMovieActors(movieDetails.id)
-                .cast
-                .map { actor -> actorMapper.map(actor) }
-
-            with(movieDetails) {
-                res.add(
-                    MovieItem(
-                        id, title, overview, posterPath, backdropPath,
-                        genres.map { genre -> GenreItem(genre.id, genre.name) },
-                        actors, if (isAdult) 18 else 0, reviews, rating, runtime
-                    )
-                )
+    override suspend fun getMovies() =
+        withContext(Dispatchers.IO) {
+            val cachedMovies = moviesLoader.loadMoviesFromCache()
+            if (cachedMovies.isEmpty()) {
+                val res = moviesLoader.loadMoviesFromNetwork()
+                save(res)
+                res.sortedByDescending { movie -> movie.rating }
+            } else {
+                cachedMovies
             }
         }
-        res
+
+    override suspend fun getMovieDetails(id: Long): MovieDetails =
```

Comment:
very strange separating classes
why u totally need it?

Reply:
I believe it's legacy, they used to be different, ok, will try to get rid off it